sdk/keyvault/keyvault-secrets/src/core/challengeBasedAuthenticationPolicy.ts

```diff
@@ -0,0 +1,138 @@
+ */
+export function challengeBasedAuthenticationPolicy(credential: TokenCredential): RequestPolicyFactory {
+  return {
+    create: (nextPolicy: RequestPolicy, options: RequestPolicyOptions) => {
+      return new ChallengeBasedAuthenticationPolicy(nextPolicy, options, credential);
+    }
+  };
+}
+
+export class AuthenticationChallenge {
+  constructor(public scopes: string[] | string) { }
+}
+
+/**
+ *
+ * Provides a RequestPolicy that can request a token from a TokenCredential
+ * implementation and then apply it to the Authorization header of a request
+ * as a Bearer token.
+ *
+ */
+export class ChallengeBasedAuthenticationPolicy extends BaseRequestPolicy {
+  private cachedToken: AccessToken | undefined = undefined;
+  private challenge: AuthenticationChallenge | undefined = undefined;
+
+  /**
+   * Creates a new ChallengeBasedAuthenticationPolicy object.
+   *
+   * @param nextPolicy The next RequestPolicy in the request pipeline.
+   * @param options Options for this RequestPolicy.
+   * @param credential The TokenCredential implementation that can supply the bearer token.
+   * @param scopes The scopes for which the bearer token applies.
+   */
+  constructor(
+    nextPolicy: RequestPolicy,
+    options: RequestPolicyOptions,
+    private credential: TokenCredential,
+    private refreshOn: number = Date.now(),
+  ) {
+    super(nextPolicy, options);
+  }
+
+  private parseWWWAuthenticate(www_authenticate: string): string {
+    let authenticateArray = www_authenticate.split(" ");
+    delete authenticateArray[0];
+    let commaSep = authenticateArray.join().split(",");
+    for (let item of commaSep) {
+      let kv = item.split("=");
+      if (kv[0].trim() == "resource") {
+        let resource = kv[1].trim().replace(/['"]+/g, '');
+        return resource;
+      }
+    }
+    return "";
+  }
+
+  /**
+   * Applies the Bearer token to the request through the Authorization header.
+   * @param webResource
+   */
+  public async sendRequest(
+    webResource: WebResource
+  ): Promise<HttpOperationResponse> {
+    if (!webResource.headers) webResource.headers = new HttpHeaders();
+
+    let originalBody = webResource.body;
+
+    if (this.challenge == undefined) {
+      webResource.body = "";
+    } else {
+      await this.authenticateRequest(webResource);
+    }
+
+    // Use a blank to start the challenge
+    let headers = webResource.headers;
+    webResource.headers = new HttpHeaders();
+    webResource.body = "";
+    let response = await this._nextPolicy.sendRequest(webResource);
+
+    if (response.status == 401) {
+      webResource.body = originalBody;
+      webResource.headers = headers;
+
+      let www_authenticate = response.headers.get("WWW-Authenticate");
+
+      if (www_authenticate) {
+        let resource = this.parseWWWAuthenticate(www_authenticate);
+        let challenge = new AuthenticationChallenge(resource + "/.default")
```

Comment:
Is this `./default` string always going to exist here? I'm curious why would the underlying resource not provide this.

Reply:
This is part of the protocol for doing the challenge.